app/app.go

```diff
@@ -0,0 +1,160 @@
+package app
+
+import (
+	"context"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+
+	dockerClient "github.com/docker/docker/client"
+	"github.com/juju/errors"
+	"github.com/projectcalico/libcalico-go/lib/apiconfig"
+	"github.com/projectcalico/libcalico-go/lib/clientv3"
+	calicov3 "github.com/projectcalico/libcalico-go/lib/clientv3"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/projecteru2/barrel/driver"
+	calicoDriver "github.com/projecteru2/barrel/driver/calico"
+	fixedIPDriver "github.com/projecteru2/barrel/driver/fixedip"
+	barrelHttp "github.com/projecteru2/barrel/http"
+	"github.com/projecteru2/barrel/proxy/docker"
+	"github.com/projecteru2/barrel/service"
+	"github.com/projecteru2/barrel/store"
+	"github.com/projecteru2/barrel/store/etcd"
+	"github.com/projecteru2/barrel/vessel"
+)
+
+// Application .
+type Application struct {
+	Mode                   string
+	DockerGID              int
+	DockerDaemonUnixSocket string
+	DockerAPIVersion       string
+	Hosts                  []string
+	DialTimeout            time.Duration
+	CertFile               string
+	KeyFile                string
+	ShutdownTimeout        time.Duration
+}
+
+// Run .
+func (app Application) Run() error {
+	if app.Mode == "default" {
+		return app.runMode(app.defaultMode)
+	} else if app.Mode == "proxy-only" {
+		return app.runMode(app.proxyOnlyMode)
+	} else if app.Mode == "network-plugin-only" {
+		return app.runMode(app.networkPluginOnlyMode)
+	}
+	return errors.New("Unrecognized barrel mode, support only [ --mode default | proxy-only | network-plugin-only ]")
+}
+
+func (app Application) runMode(serviceFactory func() ([]service.Service, error)) error {
+	sigs := make(chan os.Signal, 1)
+	signal.Notify(sigs, syscall.SIGINT, syscall.SIGHUP, syscall.SIGTERM)
+	defer signal.Stop(sigs)
+	defer close(sigs)
+
+	go func() {
```

Comment:
im not quite sure the purpose of this goroutine.

Reply:
it's legency code and removed now.

Reply:
> it's legency code and removed now.

i mean the following commit which i'll fix the review issues